torchbenchmark/_components/ncu/analyzer.py

```diff
@@ -0,0 +1,133 @@
+    It searches for the 'ncu' command in the system PATH, determines its location, and appends the
+    'extras/python' directory to the Python path.
+
+    Raises:
+        FileNotFoundError: If the 'ncu' command is not found in the system PATH.
+        FileNotFoundError: If the 'extras/python' directory does not exist in the determined NCU path.
+    """
+    ncu_path = shutil.which("ncu")
+    if not ncu_path:
+        raise FileNotFoundError("Could not find 'ncu' command in PATH.")
+    ncu_path = os.path.dirname(ncu_path)
+    if not os.path.exists(os.path.join(ncu_path, "extras/python")):
+        raise FileNotFoundError(
+            f"'extras/python' does not exist in the provided ncu_path: {ncu_path}"
+        )
+    sys.path.append(os.path.join(ncu_path, "extras/python"))
+
+
+def get_mem_traffic(kernel):
+    return (
+        kernel.metric_by_name(short_ncu_metric_name["dram_bytes_read"]).value(),
+        kernel.metric_by_name(short_ncu_metric_name["dram_bytes_write"]).value(),
+    )
+
+
+# Reference: ncu_install_path/sections/SpeedOfLight_Roofline.py
+# and ncu_install_path/sections/SpeedOfLight_RooflineChart.section
+def get_arithmetic_intensity(kernel):
+    fp32_add_achieved = kernel.metric_by_name(
+        short_ncu_metric_name["inst_executed_fadd"]
+    ).value()
+    fp32_mul_achieved = kernel.metric_by_name(
+        short_ncu_metric_name["inst_executed_fmul"]
+    ).value()
+    fp32_fma_achieved = kernel.metric_by_name(
+        short_ncu_metric_name["inst_executed_ffma"]
+    ).value()
+    fp32_achieved = fp32_add_achieved + fp32_mul_achieved + 2 * fp32_fma_achieved
+    fp64_add_achieved = kernel.metric_by_name(
+        short_ncu_metric_name["inst_executed_dadd"]
+    ).value()
+    fp64_mul_achieved = kernel.metric_by_name(
+        short_ncu_metric_name["inst_executed_dmul"]
+    ).value()
+    fp64_fma_achieved = kernel.metric_by_name(
+        short_ncu_metric_name["inst_executed_dfma"]
+    ).value()
+    fp64_achieved = fp64_add_achieved + fp64_mul_achieved + 2 * fp64_fma_achieved
+    sm_freq = kernel.metric_by_name(short_ncu_metric_name["sm_freq"]).value()
+    fp32_flops = fp32_achieved * sm_freq
+    fp64_flops = fp64_achieved * sm_freq
+    dram_bandwidth = kernel.metric_by_name(
+        short_ncu_metric_name["dram_bandwidth"]
+    ).value()
+    fp32_arithmetic_intensity = fp32_flops / dram_bandwidth
+    fp64_arithmetic_intensity = fp64_flops / dram_bandwidth
+    return fp32_arithmetic_intensity, fp64_arithmetic_intensity
+
+
+def read_ncu_report(report_path: str, required_metrics: List[str]):
+    assert os.path.exists(
+        report_path
+    ), f"The NCU report at {report_path} does not exist. Ensure you add --metrics ncu_rep to your benchmark run."
+    import_ncu_python_path()
+    import ncu_report
+
+    # save all kernels' metrics. {metric_name: [kernel1_metric_value, kernel2_metric_value, ...]}
+    results = defaultdict(list)
```

Comment:
@xuzhao9
Any suggestions on how we should save this data? We need to keep the metric results for each kernel, but we also need aggregated results, right? For example, the memory traffic (both read and write) for the whole operator should be the sum of all kernels' read and write traffic.

Reply:
@xuzhao9 @eellison 
Do you think the arithmetic intensity of the whole operator can be represented as a weighted average based on execution time?